azure/services/natgateways/natgateways.go

```diff
@@ -55,119 +57,61 @@ func New(scope NatGatewayScope) *Service {
+	var result error
 	for _, natGatewaySpec := range s.Scope.NatGatewaySpecs() {
-		existingNatGateway, err := s.getExisting(ctx, natGatewaySpec)
-
-		switch {
-		case err != nil && !azure.ResourceNotFound(err):
-			return errors.Wrapf(err, "failed to get nat gateway %s in %s", natGatewaySpec.Name, s.Scope.ResourceGroup())
-		case err == nil:
-			// nat gateway already exists
-			s.Scope.V(4).Info("nat gateway already exists", "nat gateway", natGatewaySpec.Name)
-			natGatewaySpec.Subnet.NatGateway.ID = existingNatGateway.ID
-
-			if existingNatGateway.NatGatewayIP.Name == natGatewaySpec.NatGatewayIP.Name {
-				// Skip update for Nat Gateway as it exists with expected values
-				s.Scope.V(4).Info("Nat Gateway exists with expected values, skipping update", "nat gateway", natGatewaySpec.Name)
-				natGatewaySpec.Subnet.NatGateway = *existingNatGateway
-				s.Scope.SetSubnet(natGatewaySpec.Subnet)
-				continue
-			} else {
-				s.Scope.V(2).Info("updating NAT gateway IP name to match the spec", "old name", existingNatGateway.NatGatewayIP.Name, "desired name", natGatewaySpec.NatGatewayIP.Name)
-			}
-		default:
-			// nat gateway doesn't exist but its name was specified in the subnet, let's create it
-			s.Scope.V(2).Info("nat gateway doesn't exist yet, creating it", "nat gateway", natGatewaySpec.Name)
-		}
-
-		natGatewayToCreate := network.NatGateway{
-			Location: to.StringPtr(s.Scope.Location()),
-			Sku:      &network.NatGatewaySku{Name: network.Standard},
-			NatGatewayPropertiesFormat: &network.NatGatewayPropertiesFormat{
-				PublicIPAddresses: &[]network.SubResource{
-					{
-						ID: to.StringPtr(azure.PublicIPID(s.Scope.SubscriptionID(), s.Scope.ResourceGroup(), natGatewaySpec.NatGatewayIP.Name)),
-					},
-				},
-			},
-		}
-		err = s.client.CreateOrUpdate(ctx, s.Scope.ResourceGroup(), natGatewaySpec.Name, natGatewayToCreate)
+		natGateway, err := async.CreateResource(ctx, s.Scope, s.client, natGatewaySpec, serviceName)
 		if err != nil {
-			return errors.Wrapf(err, "failed to create nat gateway %s in resource group %s", natGatewaySpec.Name, s.Scope.ResourceGroup())
+			if !azure.IsOperationNotDoneError(err) || result == nil {
+				result = err
+			}
 		}
-		s.Scope.V(2).Info("successfully created nat gateway", "nat gateway", natGatewaySpec.Name)
-		natGateway := infrav1.NatGateway{
-			ID:   azure.NatGatewayID(s.Scope.SubscriptionID(), s.Scope.ResourceGroup(), natGatewaySpec.Name),
-			Name: natGatewaySpec.Name,
-			NatGatewayIP: infrav1.PublicIPSpec{
-				Name: natGatewaySpec.NatGatewayIP.Name,
-			},
+		if err == nil && result != nil {
+			// TODO: consider making OwnerResourceName() a no-op
+			subnetSpec := s.Scope.Subnet(natGatewaySpec.OwnerResourceName())
+			networkNatGateway, ok := natGateway.(network.NatGateway)
+			if !ok {
+				// TODO: maybe this error shouldn't take priority
+				result = errors.Errorf("created resource %T is not a network.NatGateway", natGateway)
```

Comment:
In theory, this error should never happen because it should always return a NAT gateway type. That being said, if it did happen, we would want it to interrupt the loop I think because it's a fatal error and if it happened for one of the nat gateways, chances are it will happen for all of them. I would just return the error directly here.